parser/src/parser.rs

```diff
@@ -0,0 +1,523 @@
+#![allow(clippy::type_complexity)]
+
+use nom::{
+    branch::alt,
+    combinator::{map, map_res},
+    error::{Error as NomError, ErrorKind, ParseError},
+    multi::{fold_many0, many0, separated_list0},
+    sequence::{pair, tuple},
+    IResult,
+};
+
+use crate::{
+    ast::{
+        ArithmBinaryOp, Ast, BasicBlock, BinaryOp, BitBinaryOp, Call, Function, GotoInsn, Insn,
+        Literal, LogicalBinaryOp, RValue, Type, UnaryOp, Value, Variable,
+    },
+    lexer::{
+        Keyword, Lexer,
+        Token::{self, *},
+    },
+};
+
+use super::{Error, Result};
+
+/// Combinator that takes one token from the input.
+fn take1(i: &[Token]) -> IResult<&[Token], Token> {
+    let first_token = i
+        .first()
+        .ok_or_else(|| nom::Err::Error(NomError::from_error_kind(i, ErrorKind::Complete)))?;
+    Ok((&i[1..], first_token.clone()))
+}
+
+/// Combinator that checks if an input starts from one of the enumerated tokens.
+fn one_of<'a>(tokens: &'a [Token]) -> impl Fn(&'a [Token]) -> IResult<&'a [Token], Token> {
+    move |i| {
+        map_res(take1, |first_token| {
+            for token in tokens {
+                if first_token == *token {
+                    return Ok(token.clone());
+                }
+            }
+            Err(nom::Err::Error(NomError::from_error_kind(
+                i,
+                ErrorKind::OneOf,
+            )))
+        })(i)
+    }
+}
+
+/// Combinator that wraps the result of the other combinator with option,
+/// returning [`None`] if it return an recoverable error.
+fn maybe<I: Clone, O, E: ParseError<I>, F>(
+    mut parser: F,
+) -> impl FnMut(I) -> IResult<I, Option<O>, E>
+where
+    F: nom::Parser<I, O, E>,
+{
+    move |i| {
+        let i_cloned = i.clone();
+        let res = parser.parse(i);
+        match res {
+            Ok((i, o)) => Ok((i, Some(o))),
+            Err(nom::Err::Error(_)) => Ok((i_cloned, None)),
+            Err(err) => Err(err),
+        }
+    }
+}
+
+/// Parse a type.
+fn typ(i: &[Token]) -> IResult<&[Token], Type> {
+    map(
+        pair(
+            one_of(&[
+                Token::Keyword(Keyword::Void),
+                Token::Keyword(Keyword::Int),
+                Token::Keyword(Keyword::Unsigned),
+                Token::Keyword(Keyword::Bool),
+            ]),
+            fold_many0(tok(Token::Star), || 0, |s, _| s + 1),
+        ),
+        |(token, stars)| {
+            let mut typ = match token {
+                Token::Keyword(Keyword::Void) => Type::Void,
+                Token::Keyword(Keyword::Int) => Type::Int32,
+                Token::Keyword(Keyword::Unsigned) => Type::Unsigned32,
+                Token::Keyword(Keyword::Bool) => Type::Bool,
+                _ => unreachable!(),
+            };
+            for _ in 0..stars {
+                typ = Type::Ref(Box::new(typ))
+            }
+            typ
```

Comment:
Let's use `fold` here and we won't need to require `typ` to be mutable:
```suggestion
(0..stars).fold(typ, |acc, _| Type::Ref(Box::new(acc)))
```

Reply:
Applied.